### Enter your name - Your ID

# 1) Milestone Lab 01 Questions <span style="color:red"> (Total marks: 20 marks) </span>

In [6]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import pearsonr


In [7]:
df = pd.read_csv('Used Car Dataset.csv')
df

Unnamed: 0                                           car_name  \
0              0                    2017 Mercedes-Benz S-Class S400   
1              1  2020 Nissan Magnite Turbo CVT XV Premium Opt BSVI   
2              2                       2018 BMW X1 sDrive 20d xLine   
3              3                           2019 Kia Seltos GTX Plus   
4              4                    2019 Skoda Superb LK 1.8 TSI AT   
...          ...                                                ...   
1548        1548           2020 Hyundai Creta SX Opt Diesel AT BSVI   
1549        1549                 2022 Renault KWID 1.0 RXL Opt BSVI   
1550        1550                           2017 Honda WR-V i-VTEC S   
1551        1551           2018 Volkswagen Polo 1.0 MPI Comfortline   
1552        1552                        2018 Maruti Swift Dzire VXI   

     registration_year     insurance_validity fuel_type  seats  kms_driven  \
0               Jul-17          Comprehensive    Petrol      5       56000   
1               Jan-21          Comprehensive    Petrol      5       30615   
2               Sep-18          Comprehensive    Diesel      5       24000   
3               Dec-19          Comprehensive    Petrol      5       18378   
4               Aug-19          Comprehensive    Petrol      5       44900   
...                ...                    ...       ...    ...         ...   
1548            Aug-20          Comprehensive    Diesel      5       35000   
1549              2022  Third Party insurance    Petrol      5       10000   
1550            Jun-17          Comprehensive    Petrol      5       49000   
1551            May-18          Comprehensive    Petrol      5       40000   
1552            May-18          Comprehensive    Petrol      5       34756   

        ownsership transmission manufacturing_year  mileage(kmpl)  engine(cc)  \
0      First Owner    Automatic               2017           7.81      2996.0   
1      First Owner    Automatic               2020          17.40       999.0   
2      First Owner    Automatic               2018          20.68      1995.0   
3      First Owner       Manual               2019          16.50      1353.0   
4      First Owner    Automatic               2019          14.67      1798.0   
...            ...          ...                ...            ...         ...   
1548   First Owner    Automatic               2020        1493.00     11345.0   
1549        999 cc         2022     Power Steering         999.00      6706.0   
1550   First Owner       Manual               2017          17.50      1199.0   
1551  Second Owner       Manual               2018          18.78       999.0   
1552   First Owner       Manual               2018          20.85      1197.0   

      max_power(bhp)  torque(Nm)  price(in lakhs)  
0             2996.0       333.0            63.75  
1              999.0      9863.0             8.99  
2             1995.0       188.0            23.75  
3             1353.0     13808.0            13.56  
4             1798.0     17746.0            24.00  
...              ...         ...              ...  
1548         11345.0       250.0            17.41  
1549          6706.0        91.0             3.98  
1550          1199.0       887.0             5.85  
1551           999.0        75.0             4.75  
1552          1197.0      8314.0             6.10  

[1553 rows x 15 columns]

## **Q1) Select rows from 54 to 64 including the row number 64 with columns with indicies: 0, 4, 7, and 13. <span style="color:red"> (2 marks) </span>**

In [9]:
# Add your code here
df.iloc[54:65,[0,4,7,13]]


Unnamed: 0 fuel_type    ownsership  torque(Nm)
54          54    Petrol   First Owner      8850.0
55          55    Diesel   First Owner      1262.0
56          56    Petrol   First Owner       867.0
57          57       CNG   First Owner        72.0
58          58    Diesel   First Owner       200.0
59          59    Petrol   First Owner      1144.0
60          60    Petrol   First Owner     11935.0
61          61    Petrol   First Owner       831.0
62          62    Petrol   First Owner     15189.0
63          63    Diesel   First Owner     11343.0
64          64    Petrol  Second Owner     20115.0

## **Q2) Rename the 'fuel_type' column into 'fuel type' permanently. <span style="color:red"> (1.5 marks) </span>**

In [11]:
# Add your code here
df.rename(columns = {'fuel_type':'fuel type'}, inplace=True)
df


Unnamed: 0                                           car_name  \
0              0                    2017 Mercedes-Benz S-Class S400   
1              1  2020 Nissan Magnite Turbo CVT XV Premium Opt BSVI   
2              2                       2018 BMW X1 sDrive 20d xLine   
3              3                           2019 Kia Seltos GTX Plus   
4              4                    2019 Skoda Superb LK 1.8 TSI AT   
...          ...                                                ...   
1548        1548           2020 Hyundai Creta SX Opt Diesel AT BSVI   
1549        1549                 2022 Renault KWID 1.0 RXL Opt BSVI   
1550        1550                           2017 Honda WR-V i-VTEC S   
1551        1551           2018 Volkswagen Polo 1.0 MPI Comfortline   
1552        1552                        2018 Maruti Swift Dzire VXI   

     registration_year     insurance_validity fuel type  seats  kms_driven  \
0               Jul-17          Comprehensive    Petrol      5       56000   
1               Jan-21          Comprehensive    Petrol      5       30615   
2               Sep-18          Comprehensive    Diesel      5       24000   
3               Dec-19          Comprehensive    Petrol      5       18378   
4               Aug-19          Comprehensive    Petrol      5       44900   
...                ...                    ...       ...    ...         ...   
1548            Aug-20          Comprehensive    Diesel      5       35000   
1549              2022  Third Party insurance    Petrol      5       10000   
1550            Jun-17          Comprehensive    Petrol      5       49000   
1551            May-18          Comprehensive    Petrol      5       40000   
1552            May-18          Comprehensive    Petrol      5       34756   

        ownsership transmission manufacturing_year  mileage(kmpl)  engine(cc)  \
0      First Owner    Automatic               2017           7.81      2996.0   
1      First Owner    Automatic               2020          17.40       999.0   
2      First Owner    Automatic               2018          20.68      1995.0   
3      First Owner       Manual               2019          16.50      1353.0   
4      First Owner    Automatic               2019          14.67      1798.0   
...            ...          ...                ...            ...         ...   
1548   First Owner    Automatic               2020        1493.00     11345.0   
1549        999 cc         2022     Power Steering         999.00      6706.0   
1550   First Owner       Manual               2017          17.50      1199.0   
1551  Second Owner       Manual               2018          18.78       999.0   
1552   First Owner       Manual               2018          20.85      1197.0   

      max_power(bhp)  torque(Nm)  price(in lakhs)  
0             2996.0       333.0            63.75  
1              999.0      9863.0             8.99  
2             1995.0       188.0            23.75  
3             1353.0     13808.0            13.56  
4             1798.0     17746.0            24.00  
...              ...         ...              ...  
1548         11345.0       250.0            17.41  
1549          6706.0        91.0             3.98  
1550          1199.0       887.0             5.85  
1551           999.0        75.0             4.75  
1552          1197.0      8314.0             6.10  

[1553 rows x 15 columns]

 **Q3) Compute the crosstab between the 'transmission' and 'seats' columns . <span style="color:red"> (1.5 mark) </span>**

In [14]:
# Add your code here
pd.crosstab(df.transmission, df.seats)



seats                4      5      6      7      8      67000
transmission                                                 
2011                     0      3      0      0      0      0
2014                     0      5      0      0      0      0
2015                     0      1      0      0      0      0
2016                     0      1      0      0      0      0
2017                     0     28      0      0      0      0
2018                     0      1      0      1      0      0
2020                     0      0      2      0      0      0
2021                     1      0      0      1      0      0
2022                     0      2      0      0      0      0
2023                     0      2      0      0      0      0
Automatic               19    533     13    103      0      0
Manual                   2    777      3     51      2      0
Power Windows Front      0      0      0      0      0      2

In [96]:
sample = df['price(in lakhs)'].sample(500)
sample

1006    34.75
647     21.00
422     52.00
625     20.00
671      6.49
        ...  
604      4.90
1333     3.75
524      5.28
538      3.37
1482     3.84
Name: price(in lakhs), Length: 500, dtype: float64

Q4) Calculate the probabilities of each fuel type ('Petrol', 'Diesel', 'CNG', '5 Seats') for people with 'price(in lakhs)' exceeding 40. <span style="color:red"> (2 marks) </span>**

In [ ]:
from sklearn.cluster import KMeans

X = pd.merge(left=df.loc[:,df.dtypes=='int64'].drop('price(in lakhs)', axis=1), right=df.loc[:,df.dtypes=='float64'], on=df.idx).values

kmeans = KMeans(n_clusters=5, random_state=0).fit(X) 

df['price(in lakhs)'] = kmeans.labels_

df

In [70]:

df2 = df[df["price(in lakhs)"] > 40]
fuel_counts = df2['fuel type'].value_counts()
total_count = fuel_counts.sum()
prop = fuel_counts / total_count
print(fuel_counts)
print(prop)


fuel type
Petrol    95
Diesel    69
Name: count, dtype: int64
fuel type
Petrol    0.579268
Diesel    0.420732
Name: count, dtype: float64


## **Q5) Perform simple random (250), stratified random (250), and clustered sampling on the 'price(in lakhs)' column. <span style="color:red">(7 marks)**</span>

In [32]:
# Add your code here
minn = df["price(in lakhs)"].min()
maxx = df["price(in lakhs)"].max()

# Calculate the range and the interval length
diff = maxx - minn
L = diff // 5


g1 = df["price(in lakhs)"][(df["price(in lakhs)"] <= minn + L)]
g2 = df["price(in lakhs)"][(df["price(in lakhs)"] > minn + L) & (df["price(in lakhs)"] <= minn + 2 * L)]
g3 = df["price(in lakhs)"][(df["price(in lakhs)"] > minn + 2 * L) & (df["price(in lakhs)"] <= minn + 3 * L)]
g4 = df["price(in lakhs)"][(df["price(in lakhs)"] > minn + 3 * L) & (df["price(in lakhs)"] <= minn + 4 * L)]
g5 = df["price(in lakhs)"][df["price(in lakhs)"] > minn + 4 * L]



sample_size = 500
total = len(df)
sample_percentage = round(sample_size / total,3)

# Sample from each group
s1 = g1.sample(round(sample_percentage * len(g1)))
s2 = g2.sample(round(sample_percentage * len(g2)))
s3 = g3.sample(round(sample_percentage * len(g3)))
s4 = g4.sample(round(sample_percentage * len(g4)))
s5 = g5.sample(round(sample_percentage * len(g5)))

# Combine the samples
s = pd.concat([s1, s2, s3, s4, s5])



## **Q6) Compute the correlation between the'seats' and 'price(in lakhs)' columns and plot the it.** <span style="color:red">**(2 marks)**</span>

In [41]:
df.isna().sum()

Unnamed: 0            0
car_name              0
registration_year     0
insurance_validity    0
fuel type             0
seats                 0
kms_driven            0
ownsership            0
transmission          0
manufacturing_year    0
mileage(kmpl)         3
engine(cc)            3
max_power(bhp)        3
torque(Nm)            4
price(in lakhs)       0
idx                   0
dtype: int64

In [97]:
pearsonr(df['seats'], df['price(in lakhs)'])

PearsonRResult(statistic=-0.0015485983688207812, pvalue=0.9513765310304375)

## **Q7) Change the 'fuel type' column values into:**

1 for 'Petrol'.

2 for 'Diesel'.

3 for 'CNG'

4 for '5 Seats'

## **Add these changed values into a new column named 'fuel type updated'. <span style="color:red"> (4 marks) </span>**

In [87]:
##df['fuel type'] = df["fuel type updated"].map(
    #lambda x: 'Petrol' if x == 1 
               #else 'Diesel' if x == 2 
#else 'CNG' if x == 3 
               #else '5 Seats')
#df

fuel_type_mapping = {
    'Petrol': 1,
    'Diesel': 2,
    'CNG': 3,
    '5 Seats': 4
}
df['fuel type updated'] = df['fuel type'].map(fuel_type_mapping)

print(df[['fuel type', 'fuel type updated']])


     fuel type  fuel type updated
0       Petrol                  1
1       Petrol                  1
2       Diesel                  2
3       Petrol                  1
4       Petrol                  1
...        ...                ...
1548    Diesel                  2
1549    Petrol                  1
1550    Petrol                  1
1551    Petrol                  1
1552    Petrol                  1

[1553 rows x 2 columns]


# End of questions!
--------------------------------------------


# 2) Syntax from Previous labs

### 2.1) df Operations

--------------------------------------------------------

### 2.2) Plotting

--------------------------------------------------------

### 2.3) Sampling

--------------------------------------------------------

### 2.4) Binning:

--------------------------------------------------------

### 2.5) Correlation:

--------------------------------------------------------

--------------------------------------------------------

# End of previous labs syntax!